README.md
# 2048 played by AI!

![](https://nathancooperjones.com/wp-content/uploads/2020/01/2048-bot.png) 

By building the game from the ground up and using reinforcement learning (or NEAT, for now), we can teach an AI to play 2048 better than a human (or a bit worse, for now)! 


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).

# [0.2.0] - 2020-01-03
### Added
 - configuration and setup for `neat`
 - skeleton for reinforcement learning agent

# [0.1.0] - 2019-12-30
### Added
 - `Game` class to `game.py` with functions to play interactive game

# [0.0.1] - 2019-10-22
### Added
 - `requirements.txt`
 - `2048.ipynb` to `nbs`
 - Github workflow for linting and testing

# [0.0.0] - 2019-10-21
### Added
 - initial repo setup


tests/conftest.py
import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), 'fixtures'))


tests/test_game.py
import numpy as np
from twentyfortyeight.game import Game

GAME = Game()


def test_check_if_there_is_a_move_left():
    # Test that we can correctly detect when legal moves remain.
    GAME.board = np.reshape(np.arange(0, 16), (4, 4))
    assert GAME._check_if_there_is_a_move_left() is True
    GAME.board = np.reshape(np.arange(1, 17), (4, 4))
    assert GAME._check_if_there_is_a_move_left() is False
    GAME.board = np.reshape(np.ones(16), (4, 4))
    assert GAME._check_if_there_is_a_move_left() is True
    GAME.reset()
    assert GAME._check_if_there_is_a_move_left() is True


def test_shift_down_1():
    # Test shift down
    expected = np.array([[0, 0, 0, 0],
                         [0, 0, 0, 0],
                         [4, 0, 0, 0],
                         [4, 2, 2, 2]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('s')

    assert np.array_equal(GAME.board, expected)


def test_shift_down_2():
    # Test shift down special consideration #1
    expected = np.array([[0, 0, 0, 0],
                         [0, 0, 0, 0],
                         [4, 0, 0, 0],
                         [4, 2, 2, 2]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [0, 0, 0, 0],
                           [2, 0, 0, 0],
                           [4, 0, 0, 0]])
    GAME.shift('s')

    assert np.array_equal(GAME.board, expected)


def test_shift_down_3():
    # Test shift down special consideration #2
    expected = np.array([[0, 0, 0, 0],
                         [4, 0, 0, 0],
                         [4, 0, 0, 0],
                         [8, 2, 2, 2]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [4, 0, 0, 0],
                           [8, 0, 0, 0]])
    GAME.shift('s')

    assert np.array_equal(GAME.board, expected)


def test_shift_up_1():
    # Test shift up
    expected = np.array([[4, 2, 2, 2],
                         [4, 0, 0, 0],
                         [0, 0, 0, 0],
                         [0, 0, 0, 0]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('w')

    assert np.array_equal(GAME.board, expected)


def test_shift_up_2():
    # Test shift up special consideration #1
    expected = np.array([[4, 2, 2, 2],
                         [4, 0, 0, 0],
                         [0, 0, 0, 0],
                         [0, 0, 0, 0]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [0, 0, 0, 0],
                           [2, 0, 0, 0],
                           [4, 0, 0, 0]])
    GAME.shift('w')

    assert np.array_equal(GAME.board, expected)


def test_shift_up_3():
    # Test shift up special consideration #2
    expected = np.array([[4, 2, 2, 2],
                         [4, 0, 0, 0],
                         [8, 0, 0, 0],
                         [0, 0, 0, 0]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [4, 0, 0, 0],
                           [8, 0, 0, 0]])
    GAME.shift('w')

    assert np.array_equal(GAME.board, expected)


def test_shift_right_1():
    # Test shift right
    expected = np.array([[0, 0, 4, 4],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('d')

    assert np.array_equal(GAME.board, expected)


def test_shift_right_2():
    # Test shift right special consideration #1
    expected = np.array([[0, 0, 4, 4],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2]])

    GAME.board = np.array([[2, 0, 2, 4],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('d')

    assert np.array_equal(GAME.board, expected)


def test_shift_right_3():
    # Test shift right special consideration #2
    expected = np.array([[0, 4, 4, 8],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2],
                         [0, 0, 0, 2]])

    GAME.board = np.array([[2, 2, 4, 8],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('d')

    assert np.array_equal(GAME.board, expected)


def test_shift_left_1():
    # Test shift left
    expected = np.array([[4, 4, 0, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0]])

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('a')

    assert np.array_equal(GAME.board, expected)


def test_shift_left_2():
    # Test shift left special consideration #1
    expected = np.array([[4, 4, 0, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0]])

    GAME.board = np.array([[2, 0, 2, 4],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('a')

    assert np.array_equal(GAME.board, expected)


def test_shift_left_3():
    # Test shift down special consideration #2
    expected = np.array([[4, 4, 8, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0],
                         [2, 0, 0, 0]])

    GAME.board = np.array([[2, 2, 4, 8],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('a')

    assert np.array_equal(GAME.board, expected)


def test_variables_are_set():
    # Test `reset()`, `MOVE_COUNTER`, and `SCORE` values
    GAME.reset()
    assert GAME.move_counter == 0
    assert GAME.score == 0
    assert GAME.move_made is None
    assert len(GAME.board.nonzero()[0]) == 2

    GAME.board = np.array([[2, 2, 2, 2],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0],
                           [2, 0, 0, 0]])
    GAME.shift('a')
    GAME.shift('d')
    GAME.shift('w')
    GAME.shift('s')
    assert GAME.move_counter == 4
    assert GAME.score == 20
    assert GAME.move_made is True

    GAME.reset()
    assert GAME.move_counter == 0
    assert GAME.score == 0
    assert GAME.move_made is None
    assert len(GAME.board.nonzero()[0]) == 2


def test_moves_dont_count():
    GAME.reset()
    GAME.board = np.array([[2, 0, 0, 2],
                           [0, 0, 0, 0],
                           [0, 0, 0, 0],
                           [0, 0, 0, 0]])
    GAME.shift('w')
    assert GAME.move_made is False
    GAME.shift('w')
    assert GAME.move_made is False
    GAME.shift('s')
    assert GAME.move_made is True


twentyfortyeight/rl.py
import os
import random
import sys
import time

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from twentyfortyeight.game import Game


class NeuralNetwork(nn.Module):
    """TODO."""
    def __init__(self):
        super(NeuralNetwork, self).__init__()

        self.number_of_actions = 4
        self.gamma = 0.99
        self.final_epsilon = 0.0001
        self.initial_epsilon = 0.9999
        self.number_of_iterations = 2000000
        self.replay_memory_size = 10000
        self.minibatch_size = 32

        # self.layer_1 = nn.Linear(16, 64)
        # nn.init.xavier_normal_(self.layer_1.weight)
        self.layer_2 = nn.Linear(64, 128)
        nn.init.xavier_normal_(self.layer_2.weight)
        self.layer_3 = nn.Linear(128, 256)
        nn.init.xavier_normal_(self.layer_3.weight)
        self.layer_4 = nn.Linear(256, 82)
        nn.init.xavier_normal_(self.layer_4.weight)
        self.layer_5 = nn.Linear(82, 32)
        nn.init.xavier_normal_(self.layer_5.weight)
        self.layer_6 = nn.Linear(32, self.number_of_actions)
        nn.init.xavier_normal_(self.layer_6.weight)

    def forward(self, x):
        """TODO."""
        # out = F.leaky_relu(self.layer_1(x))
        # out = F.leaky_relu(self.layer_2(out))
        out = F.leaky_relu(self.layer_2(x))
        out = F.leaky_relu(self.layer_3(out))
        out = F.leaky_relu(self.layer_4(out))
        out = F.leaky_relu(self.layer_5(out))
        out = self.layer_6(out)

        return out


def train(model, start):
    """TODO."""
    # define Adam optimizer
    optimizer = optim.Adam(model.parameters(), lr=1e-6)

    # initialize mean squared error loss
    criterion = nn.MSELoss()

    # instantiate game
    game_state = Game()

    # initialize replay memory
    replay_memory = list()

    # initial action is do nothing
    action = torch.zeros([model.number_of_actions], dtype=torch.float32)
    action[2] = 1  # go down because i said so
    board, reward, terminal = game_state.step(action)
    state = torch.cat((board, board, board, board)).unsqueeze(0)

    # initialize epsilon value
    epsilon = model.initial_epsilon
    iteration = 0

    epsilon_decrements = np.linspace(model.initial_epsilon, model.final_epsilon, model.number_of_iterations)

    # main infinite loop
    while iteration < model.number_of_iterations:
        # get output from the neural network
        output = model(state)[0]

        # initialize action
        action = torch.zeros([model.number_of_actions], dtype=torch.float32)
        if torch.cuda.is_available():  # put on GPU if CUDA is available
            action = action.cuda()

        # epsilon greedy exploration
        random_action = random.random() <= epsilon
        if random_action:
            print("Performed random action!")
        action_index = [torch.randint(model.number_of_actions, torch.Size([]), dtype=torch.int)
                        if random_action
                        else torch.argmax(output)][0]

        if torch.cuda.is_available():  # put on GPU if CUDA is available
            action_index = action_index.cuda()

        action[action_index] = 1

        print(game_state.board)

        # get next state and reward
        board_1, reward, terminal = game_state.step(output)
        state_1 = torch.cat((torch.reshape(state, (4, 16))[1],
                             torch.reshape(state, (4, 16))[2],
                             torch.reshape(state, (4, 16))[3],
                             board_1)).unsqueeze(0)

        action = action.unsqueeze(0)
        reward = torch.from_numpy(np.array([reward], dtype=np.float32)).unsqueeze(0)

        # save transition to replay memory
        replay_memory.append((state, action, reward, state_1, terminal))

        # if replay memory is full, remove the oldest transition
        if len(replay_memory) > model.replay_memory_size:
            replay_memory.pop(0)

        # epsilon annealing
        epsilon = epsilon_decrements[iteration]

        # sample random minibatch
        minibatch = random.sample(replay_memory, min(len(replay_memory), model.minibatch_size))

        # unpack minibatch
        state_batch = torch.cat(tuple(d[0] for d in minibatch))
        action_batch = torch.cat(tuple(d[1] for d in minibatch))
        reward_batch = torch.cat(tuple(d[2] for d in minibatch))
        state_1_batch = torch.cat(tuple(d[3] for d in minibatch))

        if torch.cuda.is_available():  # put on GPU if CUDA is available
            state_batch = state_batch.cuda()
            action_batch = action_batch.cuda()
            reward_batch = reward_batch.cuda()
            state_1_batch = state_1_batch.cuda()

        # get output for the next state
        output_1_batch = model(state_1_batch)

        # set y_j to r_j for terminal state, otherwise to r_j + gamma*max(Q)
        y_batch = torch.cat(tuple(reward_batch[i] if minibatch[i][4]
                                  else reward_batch[i] + model.gamma * torch.max(output_1_batch[i])
                                  for i in range(len(minibatch))))

        # extract Q-value
        q_value = torch.sum(model(state_batch) * action_batch, dim=1)

        # PyTorch accumulates gradients by default, so they need to be reset in each pass
        optimizer.zero_grad()

        # returns a new Tensor, detached from the current graph, the result will never require gradient
        y_batch = y_batch.detach()

        # calculate loss
        loss = criterion(q_value, y_batch)

        # do backward pass
        loss.backward()
        optimizer.step()

        # set state to be state_1
        board = board_1
        iteration += 1

        if iteration % 25000 == 0:
            torch.save(model, "pretrained_model/current_model_" + str(iteration) + ".pth")

        print("iteration:", iteration, "elapsed time:", time.time() - start, "epsilon:", epsilon, "action:",
              action_index.cpu().detach().numpy(), "reward:", reward.numpy()[0][0], "Q max:",
              np.max(output.cpu().detach().numpy()))


def test(model):
    """TODO."""
    game_state = Game()

    # initial action is do nothing
    action = torch.zeros([model.number_of_actions], dtype=torch.float32)
    action[0] = 1
    board, reward, terminal = game_state.step(action)
    state = torch.cat((board, board, board, board)).unsqueeze(0)

    while True:
        # get output from the neural network
        print(game_state.board)

        output = model(state)[0]

        action = torch.zeros([model.number_of_actions], dtype=torch.float32)
        if torch.cuda.is_available():  # put on GPU if CUDA is available
            action = action.cuda()

        # get action
        action_index = torch.argmax(output)
        if torch.cuda.is_available():  # put on GPU if CUDA is available
            action_index = action_index.cuda()
        action[action_index] = 1

        # get next state
        board_1, reward, terminal = game_state.step(action)
        state_1 = torch.cat((torch.reshape(state, (4, 16))[1],
                             torch.reshape(state, (4, 16))[2],
                             torch.reshape(state, (4, 16))[3],
                             board_1)).unsqueeze(0)

        # set state to be state_1
        state = state_1


def main(mode):
    """TODO."""
    cuda_is_available = torch.cuda.is_available()

    if mode == 'test':
        model = torch.load(
            'pretrained_model/current_model_2000000.pth',
            map_location='cpu' if not cuda_is_available else None
        ).eval()

        if cuda_is_available:  # put on GPU if CUDA is available
            model = model.cuda()

        test(model)

    elif mode == 'train':
        if not os.path.exists('pretrained_model/'):
            os.mkdir('pretrained_model/')

        model = NeuralNetwork()

        if cuda_is_available:  # put on GPU if CUDA is available
            model = model.cuda()

        start = time.time()

        train(model, start)


if __name__ == "__main__":
    main(sys.argv[1])


twentyfortyeight/evolve.py
import os
import pickle

import neat
import numpy as np
from twentyfortyeight.game import Game
import twentyfortyeight.visualize as visualize


def eval_genomes(genomes, config):
    for genome_id, genome in genomes:
        genome.fitness = 0.0
        net = neat.nn.FeedForwardNetwork.create(genome, config)
        game = Game()

        while not game.game_over:
            output = net.activate(game.game_state())

            _, _, _ = game.step(output)

        # game._pretty_print(game.board)
        # print()

        genome.fitness = float(game.score) + (0.5 * np.max(game.board))


def run(config_file):
    # Load configuration.
    config = neat.Config(neat.DefaultGenome, neat.DefaultReproduction,
                         neat.DefaultSpeciesSet, neat.DefaultStagnation,
                         config_file)

    # Create the population, which is the top-level object for a NEAT run.
    p = neat.Population(config)

    # Add a stdout reporter to show progress in the terminal.
    p.add_reporter(neat.StdOutReporter(True))
    stats = neat.StatisticsReporter()
    p.add_reporter(stats)
    p.add_reporter(neat.Checkpointer(50))

    # Run for up to 1000 generations.
    winner = p.run(eval_genomes, 1000)
    # Save the winner.
    with open('winner-feedforward', 'wb') as f:
        pickle.dump(winner, f)

    visualize.plot_stats(stats, ylog=True, view=True, filename='feedforward-fitness.svg')
    visualize.plot_species(stats, view=True, filename='feedforward-speciation.svg')


if __name__ == '__main__':
    # Determine path to configuration file. This path manipulation is
    # here so that the script will run successfully regardless of the
    # current working directory.
    local_dir = os.path.dirname(__file__)
    config_path = os.path.join(local_dir, 'config-neat')
    run(config_path)


twentyfortyeight/test-neat.py
import os
import pickle

import neat
from twentyfortyeight.game import Game


def run_game(genome, config):
    net = neat.nn.FeedForwardNetwork.create(genome, config)
    game = Game()

    while not game.game_over:
        game._pretty_print(game.board)
        print()

        output = net.activate(game.game_state())

        _, _, _ = game.step(output)

    game._pretty_print(game.board)
    print()

    print(game.score)


def run(config_file):
    # Load configuration.
    config = neat.Config(neat.DefaultGenome, neat.DefaultReproduction,
                         neat.DefaultSpeciesSet, neat.DefaultStagnation,
                         config_file)

    genome_file = 'winner-feedforward'
    genome = pickle.load(open(genome_file, 'rb'))

    run_game(genome, config)


if __name__ == '__main__':
    # Determine path to configuration file. This path manipulation is
    # here so that the script will run successfully regardless of the
    # current working directory.
    local_dir = os.path.dirname(__file__)
    config_path = os.path.join(local_dir, 'config-neat')
    run(config_path)


twentyfortyeight/game.py
import random

from IPython.display import clear_output
import numpy as np


class Game():
    """TODO."""
    def __init__(self, size=4, probability_of_spawning_4=0.1):
        # Just so I don't get confused by this naming scheme in the future:
        # [[A, 0, 0, B],
        #  [0, 0, 0, 0],
        #  [0, 0, 0, 0],
        #  [C, 0, 0, D]]
        # A is on the upper row, left column
        # B is on the uper row, right column
        # C is on the lower row, left column
        # D is on the lower row, right column
        self.upper_row = 0
        self.left_col = 0
        self.lower_row = size - 1
        self.right_col = size - 1

        self.score = 0
        self.move_counter = 0

        self.probability_of_spawning_4 = probability_of_spawning_4

        self.move_made = None
        self.game_over = False

        self.board = np.zeros((size, size)).astype(int)
        self.num_of_board_spaces = self.board.size

        self.reset()

    def _add_random_space(self):
        """TODO."""
        list_of_zeros = list(zip(np.where(self.board == 0)[0], np.where(self.board == 0)[1]))
        if list_of_zeros:
            if random.random() < self.probability_of_spawning_4:
                self.board[random.choice(list_of_zeros)] = 4
            else:
                self.board[random.choice(list_of_zeros)] = 2
        else:
            # TODO: raise error of some sort.
            # print('You lose!')
            self.game_over = True

    def _recursively_move(self, row, col, direction, combine=False):
        """TODO."""
        # ensure we have proper checks and adjusted rows/cols for the specified direction
        adjusted_row_dict = {'s': row+1, 'w': row-1, 'd': row, 'a': row}
        adjusted_col_dict = {'s': col, 'w': col, 'd': col+1, 'a': col-1}
        adjusted_row = adjusted_row_dict[direction]
        adjusted_col = adjusted_col_dict[direction]

        if (direction == 's'
                and (row < self.upper_row or col < self.left_col
                     or row > self.lower_row - 1 or col > self.right_col)):
            return
        elif (direction == 'w'
              and (row - 1 < self.upper_row or col < self.left_col
                   or row > self.lower_row or col > self.right_col)
              ):
            return
        elif (direction == 'd'
              and (row < self.upper_row or col < self.left_col
                   or row > self.lower_row or col > self.right_col - 1)
              ):
            return
        elif (direction == 'a'
              and (row < self.upper_row or col - 1 < self.left_col
                   or row > self.lower_row or col > self.right_col)
              ):
            return

        # now general logic
        elif self.board[row, col] == 0:
            return
        elif combine is False and self.board[adjusted_row, adjusted_col] == 0:
            self.board[adjusted_row, adjusted_col] = self.board[row, col]
        elif combine is True and self.board[row, col] == self.board[adjusted_row, adjusted_col]:
            self.board[adjusted_row, adjusted_col] *= 2
            self.score += self.board[adjusted_row, adjusted_col]
        else:
            return

        self.board[row, col] = 0
        self.move_made = True
        if combine is False:
            self._recursively_move(adjusted_row, adjusted_col, direction, combine=False)

    def _can_we_move_this_block(self, row, col):
        """TODO."""
        if col > self.left_col:
            if self.board[row, col] == self.board[row, col - 1]:
                return True
        if col < self.right_col:
            if self.board[row, col] == self.board[row, col + 1]:
                return True
        if row > self.upper_row:
            if self.board[row, col] == self.board[row - 1, col]:
                return True
        if row < self.lower_row:
            if self.board[row, col] == self.board[row + 1, col]:
                return True
        else:
            return False

    def _transform_player_direction_input(self):
        """TODO."""
        raw_direction = input('Enter a direction [w/a/s/d or down/up/right/left]: ')
        direction = raw_direction.strip().lower()

        if direction == 'quit' or direction == 'q' or direction == 'exit':
            raise KeyboardInterrupt('Goodbye!')
        elif direction == 'right':
            direction = 'd'
        elif direction == 'left':
            direction = 'a'
        elif direction == 'up':
            direction = 'w'
        elif direction == 'down':
            direction = 's'
        if direction in ['w', 'a', 's', 'd']:
            return direction

        print(f'"{raw_direction}" is not valid. Try again or type "quit".')
        return self._transform_player_direction_input()

    def _pretty_print(self, mat, colors=True):
        """TODO."""
        print('\n'.join([''.join(
            [self._pretty_print_color_helper(item, colors) for item in row]) for row in mat])
        )

    def _pretty_print_color_helper(self, item, colors):
        if not colors:
            return '{:5}'.format(item)

        if item == 0:
            return '\033[97m {:5}\033[00m'.format(item)
        elif item == 2:
            return '\033[31m {:5}\033[00m'.format(item)
        elif item == 4:
            return '\033[91m {:5}\033[00m'.format(item)
        elif item == 8:
            return '\033[33m {:5}\033[00m'.format(item)
        elif item == 16:
            return '\033[34m {:5}\033[00m'.format(item)
        elif item == 32:
            return '\033[35m {:5}\033[00m'.format(item)
        elif item == 64:
            return '\033[36m {:5}\033[00m'.format(item)
        elif item == 128:
            return '\033[32m {:5}\033[00m'.format(item)
        elif item == 256:
            return '\033[92m {:5}\033[00m'.format(item)
        elif item == 512:
            return '\033[93m {:5}\033[00m'.format(item)
        elif item == 1024:
            return '\033[94m {:5}\033[00m'.format(item)
        elif item == 2048:
            return '\033[95m {:5}\033[00m'.format(item)
        elif item == 4096:
            return '\033[96m {:5}\033[00m'.format(item)
        else:
            return '\033[98m {:5}\033[00m'.format(item)

    def _check_if_there_is_a_move_left(self):
        """TODO."""
        return sum(self._what_directions_can_we_move()) > 0

    def _what_directions_can_we_move(self):
        directions_we_can_move = list()
        original_board = self.board.copy()

        for direction in ['w', 'a', 's', 'd']:
            self.shift(direction)
            directions_we_can_move.append(not np.array_equal(original_board, self.board))
            self.board = original_board.copy()

        return directions_we_can_move

    def step(self, actions):
        """TODO."""
        assert len(actions) == 4

        # Ignore invalid moves.
        directions_we_can_move = self._what_directions_can_we_move()
        viable_actions = [actions[i] for i in range(len(actions)) if directions_we_can_move[i]]
        direction = ['w', 'a', 's', 'd'][actions.index(max(viable_actions))]
        # direction = ['w', 'a', 's', 'd'][action.index(1)]

        initial_score = self.score
        self.shift(direction)

        game_board = self.game_state()
        reward = self.score - initial_score

        self._add_random_space()

        if not self._check_if_there_is_a_move_left():
            self.game_over = True

        if self.game_over:
            reward = -1
        else:
            reward += 1

        return game_board, reward, self.game_over

    def shift(self, direction):
        """TODO."""
        if self.move_made is None or self.move_made is True:
            self.move_made = False

        if direction == 's':
            # down shift everything twice, combine only once
            for iteration in range(2):
                for row_start in range(self.lower_row - 1, self.upper_row - 1, -1):
                    for row in range(row_start, self.upper_row - 1, -1):
                        for col in range(self.left_col, self.right_col + 1):
                            self._recursively_move(row, col, direction, combine=False)
                # only combine once
                if iteration == 0:
                    for row in range(self.lower_row - 1, self.upper_row - 1, -1):
                        for col in range(self.left_col, self.right_col + 1):
                            self._recursively_move(row, col, direction, combine=True)
        elif direction == 'd':
            # right shift everything twice, combine only once
            for iteration in range(2):
                for col_start in range(self.right_col - 1, self.left_col - 1, -1):
                    for col in range(col_start, self.left_col - 1, -1):
                        for row in range(self.upper_row, self.lower_row + 1):
                            self._recursively_move(row, col, direction, combine=False)
                # only combine once
                if iteration == 0:
                    for col in range(self.right_col - 1, self.left_col - 1, -1):
                        for row in range(self.upper_row, self.lower_row + 1):
                            self._recursively_move(row, col, direction, combine=True)
        elif direction == 'w':
            # up shift everything twice, combine only once
            for iteration in range(2):
                for row_start in range(self.upper_row + 1, self.lower_row + 1):
                    for row in range(row_start, self.lower_row + 1):
                        for col in range(self.left_col, self.right_col + 1):
                            self._recursively_move(row, col, direction, combine=False)
                # only combine once
                if iteration == 0:
                    for row in range(self.upper_row + 1, self.lower_row + 1):
                        for col in range(self.left_col, self.right_col + 1):
                            self._recursively_move(row, col, direction, combine=True)
        elif direction == 'a':
            # left shift everything twice, combine only once
            for iteration in range(2):
                for col_start in range(self.left_col + 1, self.right_col + 1):
                    for col in range(col_start, self.right_col + 1):
                        for row in range(self.upper_row, self.lower_row + 1):
                            self._recursively_move(row, col, direction, combine=False)
                # only combine once
                if iteration == 0:
                    for col in range(self.left_col + 1, self.right_col + 1):
                        for row in range(self.upper_row, self.lower_row + 1):
                            self._recursively_move(row, col, direction, combine=True)
        else:
            # TODO: raise error of some sort.
            print('ERROR')

        self.move_counter += 1

    def game_state(self):
        """TODO."""
        return self.board.flatten()

    def reset(self):
        """TODO."""
        self.score = 0
        self.move_counter = 0
        self.move_made = None
        self.board = np.zeros((self.lower_row + 1, self.right_col + 1)).astype(int)

        # add in some random spaces.
        for _ in range(2):
            self._add_random_space()

    def play_interactive(self, colors=True, debug=False):
        """TODO."""
        try:
            # Start by setting up the self.board.
            self.reset()

            while True:
                if debug is False:
                    clear_output()

                if self.move_made is False:
                    print("Move not made! Try again.")
                    self.move_counter -= 1

                print(f'Score: {int(self.score)}, Move Counter: {self.move_counter}')
                print('-------------------------')
                print()

                if debug is False:
                    self._pretty_print(self.board, colors)
                else:
                    print(self.board)
                print()

                # Check to make sure there is a move left we can make.
                if self._check_if_there_is_a_move_left() is False:
                    print(f'You Lose! Final self.score: {self.score}')
                    return

                # Now ask the user for input, then send that to shift.
                direction = self._transform_player_direction_input()
                self.shift(direction)
                if self.move_made:
                    # Add another random piece to the self.board.
                    self._add_random_space()

        except Exception as e:
            print(f'Uh-oh! Something went wrong! This is not good.',
                  'Here is some specifics on what went so poorly: ')
            print(e.message, e.args)
